src/services/catalog.service.ts

```diff
@@ -113,85 +112,101 @@ export class CatalogService {
-          filteredSubCats = this.filterSubCatsByTags(category.subCategories, itemTags);
+    let itemCategorized: boolean;
+    this.categories = angular.copy(this.constants.SERVICE_CATALOG_CATEGORIES);
+
+    _.each(items, (item) => {
+      itemCategorized = false;
+      _.each(this.categories, (category) => {
+        if (category.tags) {
+          if (!_.isEmpty(this.getMatchingTags(category.tags, item.tags))) {
+            itemCategorized = this.categorizeItem(item, category, 'all');
+            filteredSubCats = this.filterSubCatsByTags(category.subCategories, item.tags);
+            if (!_.isEmpty(filteredSubCats)) {
+              _.each(filteredSubCats, (subCategory) => {
+                this.categorizeItem(item, category, subCategory);
+              });
+            } else {
+              this.categorizeItem(item, category, 'other');
+            }
+          }
+        } else {
+          filteredSubCats = this.filterSubCatsByTags(category.subCategories, item.tags);
           if (!_.isEmpty(filteredSubCats)) {
-            _.each(filteredSubCats, (subCat) => {
-              this.categorize(catsBySubCats, category, subCat.id);
+            itemCategorized = this.categorizeItem(item, category, 'all');
+            _.each(filteredSubCats, (subCategory) => {
+              this.categorizeItem(item, category, subCategory);
             });
-          } else {
-            this.categorize(catsBySubCats, category, otherId);
           }
         }
-      } else {
-        filteredSubCats = this.filterSubCatsByTags(category.subCategories, itemTags);
-        if (!_.isEmpty(filteredSubCats)) {
-          _.each(filteredSubCats, (subCat) => {
-            this.categorize(catsBySubCats, category, subCat.id);
-          });
-        }
+      });  // .ea category
+      if (!itemCategorized) {
+        this.categorizeItem(item, 'other', 'all');
       }
-    });  // .ea category
+      this.categorizeItem(item, 'all', 'all');
+    });  // .ea item
+  }
 
-    if (_.isEmpty(catsBySubCats)) {
-      this.categorize(catsBySubCats, otherMainCategory, otherId);
+  private categorizeItem(item: any, category: any, subCategory: any) {
+    if (_.isString(category)) {
+      category = this.getAllOrOtherMainCategory(category);
+    }
+    if (_.isString(subCategory)) {
+      subCategory = this.getAllOrOtherSubCategory(category, subCategory);
     }
 
-    return catsBySubCats;
+    subCategory.items = _.isArray(subCategory.items) ? subCategory.items.concat([item]) : [item];
+
+    return category.hasItems = subCategory.hasItems = true;
   }
 
-  public hasCategory(item: any, category: string) {
-    let found: boolean = false;
+  private getAllOrOtherMainCategory(category: string) {
 
-    _.each(item.catsBySubCats, (cats: any) => {
-      found = _.includes(cats, category);
-      return !found;  // to break out of _.each when found
-    });
+    let categoryObj: any = _.find(this.categories, {id: category});
```

Comment:
this find is going to be run for every single item in the catalog since every item goes into the top level all, not terrible at current scales but we need to be efficient about handling thousands of items potentially. It would be better to create private variables that point to the top level "all" and "other" categories.  Also why not just always add all and other? You already have a concept of hiding/showing categories depending on whether hasItems is set.